### PR TITLE
feat: reduce lifecycle lock scope (#2242)

### PR DESCRIPTION
**Pull request by tustvold**
The future returned by `LifecyclePolicy::check_for_work` currently has a lifetime coupled to that of the mutable borrow of self. This has the unfortunate side-effect of forcing an exclusive lock to be held on the lifecycle policy whilst waiting on the returned future. 

This is despite the fact that the returned future is not actually driving the lifecycle - the lifecycle is entirely synchronous and the returned future completes when the lifecycle should be polled again.

The fix is relatively simple, and not only reduces the time the locks are held, but also removes some async from methods that don't really make sense to be async.

Relates to #2242

### CODE REVIEW

File: tracker/src/task.rs [hunk 1]
```diff
@@ -116,6 +116,58 @@ struct TrackerState {
     notify: Notify,
 }
 
+impl TrackerState {
```
Comment (tustvold) [inline tracker/src/task.rs:119]: This is just moved from being a member function on TaskTracker

File: tracker/src/task.rs [hunk 2]
```diff
@@ -385,21 +395,24 @@ where
...
-        // handles the spurious wakeup
-        loop {
-            // Request notification before checking if complete
-            // to avoid a race condition
-            let notify = self.state.notify.notified();
-
-            if self.is_complete() {
-                return;
-            }
+    pub fn join(&self) -> impl std::future::Future<Output = ()> {
```
[the remark this answers is not in the document]
Comment (alamb) [reply tracker/src/task.rs:398]: I don't know how you figured this out, but it makes sense to me

File: server/src/db.rs [hunk 3]
```diff
@@ -819,15 +819,16 @@ impl Db {
             loop {
                 self.worker_iterations_lifecycle
                     .fetch_add(1, Ordering::Relaxed);
-                let mut guard = self.lifecycle_policy.lock().await;
 
-                let policy = guard
-                    .as_mut()
-                    .expect("lifecycle policy should be initialized");
+                let fut = {
+                    let mut guard = self.lifecycle_policy.lock();
```
Comment (alamb) [inline server/src/db.rs:824]: By changing this from `async` to sync doesn't that mean the lifecycle calculations happen while holding a mutext (and nothing else can run)? I would expect that the life cycle calculations can take an extended period of time -- what else will be blocked while this is held?

File: tracker/src/task.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tustvold) [reply tracker/src/task.rs:398]: Informed trial and error mostly :laughing:

**Pull request merged**